### PR TITLE
refactor(duckdb): clean up struct packing and case statement translation

### PR DESCRIPTION
**Pull request by cpcloud**
This PR refactors `struct_pack` in the duckdb backend and deduplicates a bit of code in CASE statement translation.

### CODE REVIEW

File: ibis/backends/base/sql/alchemy/datatypes.py [hunk 1]
```diff
@@ -30,15 +30,15 @@ def compiles_array(element, compiler, **kw):
 class StructType(sat.UserDefinedType):
     def __init__(
         self,
-        pairs: Iterable[tuple[str, sat.TypeEngine]],
+        pairs: Iterable[tuple[str, sat.TypeEngine]] | Mapping[str, sat.TypeEngine],
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/datatypes.py:33]: Do we really support iterables here?

Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/datatypes.py:33]: Maybe not, I'll investigate.

File: ibis/backends/base/sql/alchemy/datatypes.py [hunk 2]
```diff
@@ -30,15 +30,15 @@ def compiles_array(element, compiler, **kw):
...
     ):
-        self.pairs = [(name, sat.to_instance(type)) for name, type in pairs]
+        self.pairs = {name: sat.to_instance(type) for name, type in dict(pairs).items()}
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/datatypes.py:35]: I find pairs a bit misleading, perhaps naming it as `fields` would be slightly better.

Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/datatypes.py:35]: Roger, fixing.

File: ibis/backends/base/sql/alchemy/datatypes.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/datatypes.py:33]: Nope, we can remove iterable support here.

File: ibis/backends/duckdb/registry.py [hunk 3]
```diff
@@ -74,36 +75,65 @@ def _timestamp_from_unix(t, op):
         raise ValueError(f"`{unit}` unit is not supported!")
 
 
+class struct_pack(GenericFunction):
+    def __init__(
+        self, mapping: Mapping[str, Any], *, type: sa.types.TypeEngine
+    ) -> None:
+        super().__init__()
+        self.mapping = mapping
```
Comment (kszucs) [inline ibis/backends/duckdb/registry.py:83]: Can we call it `values` instead?

Comment (cpcloud) [reply ibis/backends/duckdb/registry.py:83]: Sure

File: ibis/backends/duckdb/registry.py [hunk 4]
```diff
@@ -74,36 +75,65 @@ def _timestamp_from_unix(t, op):
...
-            text = sa.text(f"struct_pack({placeholders})")
-            bound_text = text.bindparams(
-                *(sa.bindparam(f"v{i:d}", val) for i, val in enumerate(value.values()))
-            )
-            name = op.name if isinstance(op, ops.Named) else "tmp"
-            params = {name: t.get_sqla_type(dtype)}
-            return bound_text.columns(**params).scalar_subquery()
+        return sa.cast(
+            sa.func.list_value(
+                *(value.tolist() if isinstance(value, np.ndarray) else value)
```
Comment (kszucs) [inline ibis/backends/duckdb/registry.py:106]: Could you please use a temporary variable for this?

Comment (cpcloud) [reply ibis/backends/duckdb/registry.py:106]: Sure :)

File: ibis/backends/duckdb/registry.py [hunk 5]
```diff
@@ -74,36 +75,65 @@ def _timestamp_from_unix(t, op):
...
+            ),
+            sqla_type,
+        )
+    elif dtype.is_floating() and not np.isfinite(value):
```
Comment (kszucs) [inline ibis/backends/duckdb/registry.py:110]: Could you move `np.isfinite()` check inside to handle all floating literals here (returning for finite values too)?

File: ibis/backends/duckdb/registry.py [hunk 6]
```diff
@@ -74,36 +75,65 @@ def _timestamp_from_unix(t, op):
...
+            type=sqla_type,
+        )
+    elif dtype.is_map():
         raise NotImplementedError(
             f"Ibis dtype `{dtype}` with mapping type "
             f"`{type(value).__name__}` isn't yet supported with the duckdb "
             "backend"
         )
-    return sa.cast(sa.literal(value), sqla_type)
+
```
Comment (kszucs) [inline ibis/backends/duckdb/registry.py:132]: Could you put an `else` branch here?

Comment (cpcloud) [reply ibis/backends/duckdb/registry.py:132]: Any particular reason?

File: ibis/backends/duckdb/registry.py (hunk 5 above)
Comment (cpcloud) [reply ibis/backends/duckdb/registry.py:110]: Possibly, let me check.

File: ibis/backends/duckdb/registry.py (hunk 6 above)
Comment (kszucs) [reply ibis/backends/duckdb/registry.py:132]: That way all branches either return with the translated value or raise so we don't need to keep in mind which datatypes can reach the end of this function.

**Review by kszucs — APPROVED** [review]
Just nits, otherwise it looks good to me!

File: ibis/backends/duckdb/registry.py (hunk 6 above)
Comment (cpcloud) [reply ibis/backends/duckdb/registry.py]: Got it, makes sense!

Comment (cpcloud) [reply ibis/backends/duckdb/registry.py]: Fixed.

File: ibis/backends/duckdb/registry.py (hunk 5 above)
Comment (cpcloud) [reply ibis/backends/duckdb/registry.py]: Done.

File: ibis/backends/duckdb/registry.py (hunk 4 above)
Comment (cpcloud) [reply ibis/backends/duckdb/registry.py]: Done.

File: ibis/backends/duckdb/registry.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/duckdb/registry.py]: Done.

File: ibis/backends/base/sql/alchemy/datatypes.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/datatypes.py]: Done.

File: ibis/backends/tests/snapshots/test_sql/test_group_by_has_index/duckdb/out.sql [hunk 7]
```diff
@@ -1,34 +1,20 @@
...
-    WHEN (
-      t0.continent = CAST('OC' AS TEXT)
-    )
-    THEN CAST('Oceania' AS TEXT)
-    WHEN (
-      t0.continent = CAST('AN' AS TEXT)
-    )
-    THEN CAST('Antarctica' AS TEXT)
-    ELSE CAST('Unknown continent' AS TEXT)
+  CASE t0.continent
```
Comment (kszucs) [inline ibis/backends/tests/snapshots/test_sql/test_group_by_has_index/duckdb/out.sql:2]: Much nicer!

**Pull request merged**